llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -618,11 +619,40 @@ class YkIRWriter {
 
   void serialiseBlock(BasicBlock &BB, ValueLoweringMap &VLMap,
                       unsigned &BBIdx) {
+    // Keep the instruction skipping logic in one place.
+    auto ShouldSkipInstr = [](Instruction *I) {
+      // Skip non-semantic instrucitons for now.
+      //
+      // We may come back to them later if we need better debugging
+      // facilities, but for now they just clutter up our AOT module.
+      return I->isDebugOrPseudoInst();
+    };
+
+    // Count instructions.
+    //
+    // FIXME: I don't like this much:
+    //
+    //  - Assumes one LLVM instruction becomes exactly one Yk IR instruction.
```

Comment:
This seems like a dangerous assumption!

Reply:
It's annoying. I agree.

The problem is, LLVM's output streamer can only do exactly that: stream. We need to be able to emit an instruction count, and then the payload for the instructions. But at present we can't say how many instructions we will emit, without actually doing the lowering.

It just so happens that (aside from debug instructions now) it's currently a 1-1 mapping...

Reply:
Why emit a count at all?

Reply:
It's because the IR will be streamed in later, and the reader needs to know how many to read. Such is serialisation.

We could use some kind of sentinel instead I suppose. If we steal an opcode, we could use that for "end of stream".

Reply:
If we don't know the size, then we can a) just buffer it until we do know the size b) put a dummy size and patch it later. Failing that an "EOS" thingy might be a good idea.

Reply:
Yes, just its just a larger change than I want to make now.

Reply:
(An llvm output streamer can't be rewound, so patching is probably out of the question)

Reply:
As a quick hack can we add a count of both LLVM and our IR and just assert at the end that the two are the same? If we ever mess things up, it'll remind us.

Reply:
https://github.com/ykjit/ykllvm/pull/124/commits/4a22f6f11d9c163ee81cf2acfb4ab0cde6135b20 kills the debug print I left in.